### PR TITLE
[Core] Tri Layer Keys

### PR DESCRIPTION
**Pull request by drashna**
## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: docs/feature_tri_layer.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+| `QK_TRI_LAYER_UPPER` | `TL_UPPR` | Momentraily enables the "upper" layer. Enables the "adjust" layer if the "lower" layer is also enabled" |
+
+## Configuration
+
+To change the default values for the layers, you can change these defines, in your `config.h`
+
+| Config name              | Default | Description                              |
+|--------------------------|---------|------------------------------------------|
+| `TRI_LAYER_LOWER_LAYER`  | `1`     | Sets the default for the "lower" layer.  |
+| `TRI_LAYER_LOWER_UPPER`  | `2`     | Sets the default for the "upper" layer.  |
```
Comment (elpekenin) [inline docs/feature_tri_layer.md:27]: A couple typos
```suggestion
| `TRI_LAYER_UPPER_LAYER`  | `2`     | Sets the default for the "upper" layer.  |
```

File: docs/feature_tri_layer.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+| `TRI_LAYER_LOWER_ADJUST` | `3`     | Sets the default for the "adjust" layer. |
```
Comment (elpekenin) [inline docs/feature_tri_layer.md:28]:
```suggestion
| `TRI_LAYER_ADJUST_LAYER` | `3`     | Sets the default for the "adjust" layer. |
```

File: docs/feature_tri_layer.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+Eg, if you wanted to set the "Adjust" layer to be layer 5, you'd add this to your `config.h`: 
+
+```c
+#define TRI_LAYER_LOWER_ADJUST 5
```
Comment (elpekenin) [inline docs/feature_tri_layer.md:33]:
```suggestion
#define TRI_LAYER_ADJUST_LAYER 5
```

File: docs/feature_tri_layer.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+Note that the "upper", "lower" and "adjust" names don't have a particular significance.  They are just used to identify and clarify the behavior, and the layers don't have to be in any particular order.  However, layers are processed from highest numeric value to lowest.
+
+For a detailed explanation of how the layer stack works, checkout [Keymap Overview](keymap.md#keymap-and-layers).
+
+## Keycodes :id=keycodes
+
+| Keycode              | Alias     | Description                                                                                             |
+|----------------------|-----------|---------------------------------------------------------------------------------------------------------|
+| `QK_TRI_LAYER_LOWER` | `TL_LOWR` | Momentraily enables the "lower" layer. Enables the "adjust" layer if the "upper" layer is also enabled" |
+| `QK_TRI_LAYER_UPPER` | `TL_UPPR` | Momentraily enables the "upper" layer. Enables the "adjust" layer if the "lower" layer is also enabled" |
```
Comment (drashna) [inline docs/feature_tri_layer.md:18]:
```suggestion
| `QK_TRI_LAYER_LOWER` | `TL_LOWR` | Momentarily enables the "lower" layer. Enables the "adjust" layer if the "upper" layer is also enabled" |
| `QK_TRI_LAYER_UPPER` | `TL_UPPR` | Momentarily enables the "upper" layer. Enables the "adjust" layer if the "lower" layer is also enabled" |
```

File: docs/feature_tri_layer.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+This enables support for the OLKB style "Tri Layer" keycodes.  These function similar to the `MO` (momentary) function key, but if both the "Lower" and "Upper" keys are pressed, it activates a third "Adjust" layer.  To enable this functionality, add this line to your `rules.mk`:
+
+```make
+TRI_LAYER_ENABLE = yes
+```
+
+Note that the "upper", "lower" and "adjust" names don't have a particular significance.  They are just used to identify and clarify the behavior, and the layers don't have to be in any particular order.  However, layers are processed from highest numeric value to lowest.
+
+For a detailed explanation of how the layer stack works, checkout [Keymap Overview](keymap.md#keymap-and-layers).
```
Comment (zvecr) [inline docs/feature_tri_layer.md:11]:
```suggestion
For a detailed explanation of how the layer stack works, check out [Keymap Overview](keymap.md#keymap-and-layers).
```

File: docs/feature_tri_layer.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
+# Tri Layers :id=tri-layers
+
+This enables support for the OLKB style "Tri Layer" keycodes.  These function similar to the `MO` (momentary) function key, but if both the "Lower" and "Upper" keys are pressed, it activates a third "Adjust" layer.  To enable this functionality, add this line to your `rules.mk`:
+
+```make
+TRI_LAYER_ENABLE = yes
+```
+
+Note that the "upper", "lower" and "adjust" names don't have a particular significance.  They are just used to identify and clarify the behavior, and the layers don't have to be in any particular order.  However, layers are processed from highest numeric value to lowest.
```
Comment (zvecr) [inline docs/feature_tri_layer.md:9]: The 2 last sentences kind of contradict each other, something more like
```suggestion
Note that the "upper", "lower" and "adjust" names don't have a particular significance, they are just used to identify and clarify the behavior. Layers are processed from highest numeric value to lowest, however the values are not required to be consecutive.
```

Comment (drashna) [reply docs/feature_tri_layer.md]: Yeah, that's a lot less ambiguous. Thanks

**Review by zvecr — APPROVED** [review]

File: quantum/process_keycode/process_tri_layer.c [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        case QK_TRI_LAYER_LOWER:
+            if (record->event.pressed) {
+                layer_on(get_tri_layer_lower_layer());
+                update_tri_layer(get_tri_layer_lower_layer(), get_tri_layer_upper_layer(), get_tri_layer_adjust_layer());
+            } else {
+                layer_off(get_tri_layer_lower_layer());
+                update_tri_layer(get_tri_layer_lower_layer(), get_tri_layer_upper_layer(), get_tri_layer_adjust_layer());
+            }
+            return false;
+            break;
```
Comment (tzarc) [inline quantum/process_keycode/process_tri_layer.c:19]:
```suggestion
```

File: quantum/process_keycode/process_tri_layer.c [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        case QK_TRI_LAYER_UPPER:
+            if (record->event.pressed) {
+                layer_on(get_tri_layer_upper_layer());
+                update_tri_layer(get_tri_layer_lower_layer(), get_tri_layer_upper_layer(), get_tri_layer_adjust_layer());
+            } else {
+                layer_off(get_tri_layer_upper_layer());
+                update_tri_layer(get_tri_layer_lower_layer(), get_tri_layer_upper_layer(), get_tri_layer_adjust_layer());
+            }
+            return false;
+            break;
```
Comment (tzarc) [inline quantum/process_keycode/process_tri_layer.c:29]:
```suggestion
```

File: quantum/tri_layer.h [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+#endif
+#ifndef TRI_LAYER_UPPER_LAYER
+#    define TRI_LAYER_UPPER_LAYER 2
+#endif
+#ifndef TRI_LAYER_ADJUST_LAYER
+#    define TRI_LAYER_ADJUST_LAYER 3
+#endif
+
+/**
+ * @brief Set the tri layer lower layer layer value
```
Comment (tzarc) [inline quantum/tri_layer.h:17]:
```suggestion
 * @brief Set the tri layer lower layer index
```

File: quantum/tri_layer.h [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * @param layer
+ */
+void set_tri_layer_lower_layer(uint8_t layer);
+/**
+ * @brief Set the tri layer upper layer layer value
```
Comment (tzarc) [inline quantum/tri_layer.h:23]:
```suggestion
 * @brief Set the tri layer upper layer index
```

File: quantum/tri_layer.h [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * @param layer
+ */
+void set_tri_layer_upper_layer(uint8_t layer);
+/**
+ * @brief Set the tri layer adjust layer object
```
Comment (tzarc) [inline quantum/tri_layer.h:29]:
```suggestion
 * @brief Set the tri layer adjust layer index
```

File: quantum/tri_layer.h [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * @param layer
+ */
+void set_tri_layer_adjust_layer(uint8_t layer);
+/**
+ * @brief Set the tri layer layers object
```
Comment (tzarc) [inline quantum/tri_layer.h:35]:
```suggestion
 * @brief Set the tri layer indices
```

File: quantum/tri_layer.h [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * @param lower
+ * @param upper
+ * @param adjust
+ */
+void set_tri_layer_layers(uint8_t lower, uint8_t upper, uint8_t adjust);
+/**
+ * @brief Get the tri layer lower layer object
```
Comment (tzarc) [inline quantum/tri_layer.h:43]:
```suggestion
 * @brief Get the tri layer lower layer index
```

File: quantum/tri_layer.h [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * @return uint8_t
+ */
+uint8_t get_tri_layer_lower_layer(void);
+/**
+ * @brief Get the tri layer upper layer object
```
Comment (tzarc) [inline quantum/tri_layer.h:49]:
```suggestion
 * @brief Get the tri layer upper layer index
```

File: quantum/tri_layer.h [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * @return uint8_t
+ */
+uint8_t get_tri_layer_upper_layer(void);
+/**
+ * @brief Get the tri layer adjust layer object
```
Comment (tzarc) [inline quantum/tri_layer.h:55]:
```suggestion
 * @brief Get the tri layer adjust layer index
```

**Review by tzarc — CHANGES_REQUESTED** [review]

**Review by tzarc — APPROVED** [review]

**Pull request merged**